kv/kvserver: TestRaftPreVote failed

kv/kvserver.TestRaftPreVote [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13478405?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13478405?buildTab=artifacts#/) on master @ [05d1395c556dfb8f6bac6d6dfcc7ae60824edd07](https://github.com/cockroachdb/cockroach/commits/05d1395c556dfb8f6bac6d6dfcc7ae60824edd07):


```
=== RUN   TestRaftPreVote
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/da89cade6788d04c28915439cd9b25f2/logTestRaftPreVote256505961
    test_log_scope.go:81: use -show-logs to present logs inline
    client_raft_test.go:6302: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/da89cade6788d04c28915439cd9b25f2/logTestRaftPreVote256505961
--- FAIL: TestRaftPreVote (42.27s)
=== RUN   TestRaftPreVote/partial=false
    --- FAIL: TestRaftPreVote/partial=false (20.33s)
=== RUN   TestRaftPreVote/partial=false/quiesce=false
        --- FAIL: TestRaftPreVote/partial=false/quiesce=false (11.41s)
=== RUN   TestRaftPreVote/partial=false/quiesce=false/expLease=false
    client_raft_test.go:6207: n1 has lease
    client_raft_test.go:6215: n3 not quiesced
    client_raft_test.go:6220: n3 partitioned
    client_raft_test.go:6225: n1 StateLeader at term=6 commit=22
    client_raft_test.go:6233: n3 not quiesced
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6137: r68 proposal rejected: GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00")
    client_raft_test.go:6239: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6242: n3 became pre-candidate
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6252: n3 is still pre-candidate
    client_raft_test.go:6259: n1 StateLeader at term=6 commit=23
    client_raft_test.go:6263: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:6263
        	            				github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:28
        	Error:      	Not equal: 
        	            	expected: 0x16
        	            	actual  : 0x17
        	Test:       	TestRaftPreVote/partial=false/quiesce=false/expLease=false
            --- FAIL: TestRaftPreVote/partial=false/quiesce=false/expLease=false (5.65s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRaftPreVote.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35259

kv/kvserver.TestRaftPreVote [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13583561?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13583561?buildTab=artifacts#/) on master @ [9b10f083f8889cdc1bc549d9428f06fd784ba081](https://github.com/cockroachdb/cockroach/commits/9b10f083f8889cdc1bc549d9428f06fd784ba081):


```
=== RUN   TestRaftPreVote
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/da89cade6788d04c28915439cd9b25f2/logTestRaftPreVote2837763692
    test_log_scope.go:81: use -show-logs to present logs inline
    client_raft_test.go:6302: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/da89cade6788d04c28915439cd9b25f2/logTestRaftPreVote2837763692
--- FAIL: TestRaftPreVote (45.69s)
=== RUN   TestRaftPreVote/partial=true/quiesce=false
        --- FAIL: TestRaftPreVote/partial=true/quiesce=false (13.48s)
=== RUN   TestRaftPreVote/partial=true/quiesce=false/expLease=false
    client_raft_test.go:6207: n1 has lease
    client_raft_test.go:6215: n3 not quiesced
    client_raft_test.go:6220: n3 partitioned
    client_raft_test.go:6225: n1 StateLeader at term=6 commit=22
    client_raft_test.go:6233: n3 not quiesced
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6137: r68 proposal rejected: GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00")
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6239: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6242: n3 became pre-candidate
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6252: n3 is still pre-candidate
    client_raft_test.go:6259: n1 StateLeader at term=6 commit=23
    client_raft_test.go:6263: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:6263
        	            				github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:28
        	Error:      	Not equal: 
        	            	expected: 0x16
        	            	actual  : 0x17
        	Test:       	TestRaftPreVote/partial=true/quiesce=false/expLease=false
            --- FAIL: TestRaftPreVote/partial=true/quiesce=false/expLease=false (7.01s)
=== RUN   TestRaftPreVote/partial=true
    --- FAIL: TestRaftPreVote/partial=true (22.86s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRaftPreVote.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestRaftPreVote [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13655700?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13655700?buildTab=artifacts#/) on master @ [d9e82c261e952ccbcc0b9f9f5df2b8e08d227339](https://github.com/cockroachdb/cockroach/commits/d9e82c261e952ccbcc0b9f9f5df2b8e08d227339):


```
=== RUN   TestRaftPreVote
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/da89cade6788d04c28915439cd9b25f2/logTestRaftPreVote190096482
    test_log_scope.go:81: use -show-logs to present logs inline
    client_raft_test.go:6302: -- test log scope end --
--- FAIL: TestRaftPreVote (45.56s)
=== RUN   TestRaftPreVote/partial=true/quiesce=false
        --- FAIL: TestRaftPreVote/partial=true/quiesce=false (14.45s)
=== RUN   TestRaftPreVote/partial=true/quiesce=false/expLease=false
    client_raft_test.go:6207: n1 has lease
    client_raft_test.go:6215: n3 not quiesced
    client_raft_test.go:6220: n3 partitioned
    client_raft_test.go:6225: n1 StateLeader at term=6 commit=22
    client_raft_test.go:6233: n3 not quiesced
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6137: r68 proposal rejected: GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00")
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6239: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6242: n3 became pre-candidate
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6252: n3 is still pre-candidate
    client_raft_test.go:6259: n1 StateLeader at term=6 commit=23
    client_raft_test.go:6263: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:6263
        	            				github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:28
        	Error:      	Not equal: 
        	            	expected: 0x16
        	            	actual  : 0x17
        	Test:       	TestRaftPreVote/partial=true/quiesce=false/expLease=false
            --- FAIL: TestRaftPreVote/partial=true/quiesce=false/expLease=false (6.95s)
=== RUN   TestRaftPreVote/partial=true
    --- FAIL: TestRaftPreVote/partial=true (23.50s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRaftPreVote.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestRaftPreVote [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/13673341?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/13673341?buildTab=artifacts#/) on master @ [e0fbbe4c1334580bf05d6d6fb0d406a327e43944](https://github.com/cockroachdb/cockroach/commits/e0fbbe4c1334580bf05d6d6fb0d406a327e43944):


```
=== RUN   TestRaftPreVote
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/c221e3b9270a405eed2e7fc80674a38f/logTestRaftPreVote461643129
    test_log_scope.go:81: use -show-logs to present logs inline
    client_raft_test.go:6302: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c221e3b9270a405eed2e7fc80674a38f/logTestRaftPreVote461643129
--- FAIL: TestRaftPreVote (43.13s)
=== RUN   TestRaftPreVote/partial=false/quiesce=false
        --- FAIL: TestRaftPreVote/partial=false/quiesce=false (12.11s)
=== RUN   TestRaftPreVote/partial=false/quiesce=false/expLease=true
    client_raft_test.go:6207: n1 has lease
    client_raft_test.go:6215: n3 not quiesced
    client_raft_test.go:6220: n3 partitioned
    client_raft_test.go:6225: n1 StateLeader at term=6 commit=22
    client_raft_test.go:6233: n3 not quiesced
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6137: r68 proposal rejected: GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00")
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6239: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6242: n3 became pre-candidate
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6252: n3 is still pre-candidate
    client_raft_test.go:6259: n1 StateLeader at term=6 commit=23
    client_raft_test.go:6263: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:6263
        	            				github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:28
        	Error:      	Not equal: 
        	            	expected: 0x16
        	            	actual  : 0x17
        	Test:       	TestRaftPreVote/partial=false/quiesce=false/expLease=true
            --- FAIL: TestRaftPreVote/partial=false/quiesce=false/expLease=true (6.20s)
=== RUN   TestRaftPreVote/partial=false
    --- FAIL: TestRaftPreVote/partial=false (21.13s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRaftPreVote.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

failed here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/13681574?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true

```
Failed
=== RUN   TestRaftPreVote/partial=false/quiesce=false/expLease=true
    client_raft_test.go:6207: n1 has lease
    client_raft_test.go:6215: n3 not quiesced
    client_raft_test.go:6220: n3 partitioned
    client_raft_test.go:6225: n1 StateLeader at term=6 commit=22
    client_raft_test.go:6233: n3 not quiesced
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6137: r68 proposal rejected: GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00")
    client_raft_test.go:6239: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6242: n3 became pre-candidate
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6252: n3 is still pre-candidate
    client_raft_test.go:6259: n1 StateLeader at term=6 commit=23
    client_raft_test.go:6263:
          Error Trace:  github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:6263
                              github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:28
          Error:        Not equal:
                        expected: 0x16
                        actual  : 0x17
          Test:         TestRaftPreVote/partial=false/quiesce=false/expLease=true
            --- FAIL: TestRaftPreVote/partial=false/quiesce=false/expLease=true (5.48s)
```

kv/kvserver.TestRaftPreVote [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13744759?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13744759?buildTab=artifacts#/) on master @ [826569e4e2e17326c83b2b92188f4eb3be91e42c](https://github.com/cockroachdb/cockroach/commits/826569e4e2e17326c83b2b92188f4eb3be91e42c):


```
=== RUN   TestRaftPreVote
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/da89cade6788d04c28915439cd9b25f2/logTestRaftPreVote3257750620
    test_log_scope.go:81: use -show-logs to present logs inline
    client_raft_test.go:6302: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/da89cade6788d04c28915439cd9b25f2/logTestRaftPreVote3257750620
--- FAIL: TestRaftPreVote (43.29s)
=== RUN   TestRaftPreVote/partial=true/quiesce=false
        --- FAIL: TestRaftPreVote/partial=true/quiesce=false (13.54s)
=== RUN   TestRaftPreVote/partial=true/quiesce=false/expLease=true
    client_raft_test.go:6207: n1 has lease
    client_raft_test.go:6215: n3 not quiesced
    client_raft_test.go:6220: n3 partitioned
    client_raft_test.go:6225: n1 StateLeader at term=6 commit=22
    client_raft_test.go:6233: n3 not quiesced
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6137: r68 proposal rejected: GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00")
    client_raft_test.go:6239: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6239: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6242: n3 became pre-candidate
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6249: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6252: n3 is still pre-candidate
    client_raft_test.go:6259: n1 StateLeader at term=6 commit=23
    client_raft_test.go:6263: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:6263
        	            				github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:28
        	Error:      	Not equal: 
        	            	expected: 0x16
        	            	actual  : 0x17
        	Test:       	TestRaftPreVote/partial=true/quiesce=false/expLease=true
            --- FAIL: TestRaftPreVote/partial=true/quiesce=false/expLease=true (6.50s)
=== RUN   TestRaftPreVote/partial=true
    --- FAIL: TestRaftPreVote/partial=true (22.16s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRaftPreVote.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -6130,10 +6130,10 @@ func TestRaftPreVote(t *testing.T) {
 				// We install a proposal filter which rejects proposals to the range
 				// during the partition (typically txn record cleanup via GC requests,
 				// but also e.g. lease extensions).
-				var partitioned atomic.Bool
+				var partitioned, blocked atomic.Bool
 				var rangeID roachpb.RangeID
 				propFilter := func(args kvserverbase.ProposalFilterArgs) *kvpb.Error {
-					if partitioned.Load() && args.Req.RangeID == rangeID {
+					if blocked.Load() && args.Req.RangeID == rangeID {
 						t.Logf("r%d proposal rejected: %s", rangeID, args.Req)
 						return kvpb.NewError(errors.New("rejected"))
 					}
@@ -6206,13 +6206,45 @@ func TestRaftPreVote(t *testing.T) {
 				tc.WaitForValues(t, key, []int64{2, 2, 2})
 				t.Logf("n1 has lease")
 
-				// Wait for the range to quiesce, if enabled.
+				// Block new proposals to the range.
+				blocked.Store(true)
+				t.Logf("n1 proposals blocked")
+
+				// Wait for the range to quiesce, if enabled. Otherwise, wait for the
+				// range to stabilize such that the leader's log does not change for a
+				// second, and has been replicated to all followers.
 				if quiesce {
 					require.Eventually(t, repl3.IsQuiescent, 10*time.Second, 100*time.Millisecond)
 					t.Logf("n3 quiesced")
 				} else {
 					require.False(t, repl3.IsQuiescent())
 					t.Logf("n3 not quiesced")
+
+					var lastIndex uint64
+					var lastChanged time.Time
+					require.Eventually(t, func() bool {
+						status := repl1.RaftStatus()
+						require.Equal(t, raft.StateLeader, status.RaftState)
+						if i := status.Progress[1].Match; i > lastIndex {
+							t.Logf("n1 last index changed: %d -> %d", lastIndex, i)
+							lastIndex, lastChanged = i, time.Now()
+							return false
+						}
+						for i, pr := range status.Progress {
+							if pr.Match != lastIndex {
+								t.Logf("n%d match %d not at n1 last index %d, waiting", i, pr.Match, lastIndex)
+								return false
+							}
+						}
+						if since := time.Since(lastChanged); since < time.Second {
+							t.Logf("n1 last index %d changed %s ago, waiting",
+								lastIndex, since.Truncate(time.Millisecond))
+							return false
+						}
+						return true
+					}, 10*time.Second, 200*time.Millisecond)
+					t.Logf("n1 stabilized range")
+					logStatus(repl1.RaftStatus())
 				}
 
 				// Partition n3.
@@ -6266,7 +6298,9 @@ func TestRaftPreVote(t *testing.T) {
 				}
 				t.Logf("n1 is still leader")
 
-				// Heal the partition, and wait for the replica to become a follower.
+				// Heal the partition and unblock proposals, then wait for the replica
+				// to become a follower.
+				blocked.Store(false)
 				partitioned.Store(false)
 				t.Logf("n3 partition healed")
 

